test/transaction_test.dart

```diff
@@ -125,62 +123,28 @@ void main() {
     });
 
     test('May intentionally rollback transaction', () async {
-      var reached = false;
-      await conn.transaction((c) async {
-        await c.query('INSERT INTO t (id) VALUES (1)');
-        c.cancelTransaction();
+      final rollback = Exception();
 
-        reached = true;
-        await c.query('INSERT INTO t (id) VALUES (2)');
-      });
-
-      expect(reached, false);
-      final result = await conn.query('SELECT id FROM t');
-      expect(result, []);
-    });
+      await expectLater(conn.runTx((c) async {
+        await c.execute('INSERT INTO t (id) VALUES (1)');
+        throw rollback;
+      }), throwsA(rollback));
 
-    test('Intentional rollback on non-transaction has no impact', () async {
-      conn.cancelTransaction();
-      final result = await conn.query('SELECT id FROM t');
+      final result = await conn.execute('SELECT id FROM t');
       expect(result, []);
     });
 
-    test('Intentional rollback from outside of a transaction has no impact',
-        () async {
-      final orderEnsurer = [];
-      final nextCompleter = Completer.sync();
-      final outResult = conn.transaction((c) async {
-        orderEnsurer.add(1);
-        await c.query('INSERT INTO t (id) VALUES (1)');
-        orderEnsurer.add(2);
-        nextCompleter.complete();
-        final result = await c.query('SELECT id FROM t');
-        orderEnsurer.add(3);
-
-        return result;
-      });
-
-      await nextCompleter.future;
-      conn.cancelTransaction();
-
-      orderEnsurer.add(11);
-      final results = await outResult;
-
-      expect(orderEnsurer, [1, 2, 11, 3]);
-      expect(results, [
-        [1]
-      ]);
-    });
-
     test('A transaction does not preempt pending queries', () async {
       // Add a few insert queries but don't await, then do a transaction that does a fetch,
       // make sure that transaction contains all of the elements.
-      conn.execute('INSERT INTO t (id) VALUES (1)');
-      conn.execute('INSERT INTO t (id) VALUES (2)');
-      conn.execute('INSERT INTO t (id) VALUES (3)');
-
-      final results = await conn.transaction((ctx) async {
-        return await ctx.query('SELECT id FROM t');
+      // This makes use of the implementation detail that simple executes
+      // (no result rows, no parameters) are locking the connection directly.
+      conn.execute('INSERT INTO t (id) VALUES (1)', ignoreRows: true);
```

Comment:
In the v3 API, only simple queries can reliably be unawaited. Since the default is to prepare a transaction and run it, there are two roundtrips involved with the extended query mode. These tests all expect `execute` to lock the connection FSM once though.

We're essentially testing implementation internals here since not awaiting queries is not something we officially support in the v3 API. But it's still very useful to assert that not awaiting statement doesn't bring the whole connection into an unusable state.